### DISCUSSION TITLE
Program plugin not visible in Ghidra

### DISCUSSION THREAD

**Discussion by Pyrdacor** [NationalSecurityAgency/ghidra#5175] [Q&A] [answered]
Hi I have created a basic program plugin through the eclipse plugin GhidraDev. It created all kind of plugin classes but I only kept the one for the program plugin. The project is linked to Ghidra and I can open Ghidra through Run or Debug in Eclipse. However I don't see any toolbar or menu entry in Ghidra. The resources on plugins are very limited on the web but looking at the examples it should work.

```java
//@formatter:off
@PluginInfo(
	status = PluginStatus.RELEASED,
	packageName = ExamplesPluginPackage.NAME,
	category = PluginCategoryNames.EXAMPLES,
	shortDescription = "Plugin short description goes here.",
	description = "Plugin long description goes here.",
	servicesRequired = { ProgramManager.class }
)
//@formatter:on
public class SymbolTransferPlugin extends ProgramPlugin {

	MyProvider provider;

	/**
	 * Plugin constructor.
	 * 
	 * @param tool The plugin tool that this plugin is added to.
	 */
	public SymbolTransferPlugin(PluginTool tool) {
		super(tool);

		// TODO: Customize provider (or remove if a provider is not desired)
		String pluginName = getName();
		provider = new MyProvider(this, pluginName);

		// TODO: Customize help (or remove if help is not desired)
		String topicName = this.getClass().getPackage().getName();
		String anchorName = "HelpAnchor";
		provider.setHelpLocation(new HelpLocation(topicName, anchorName));
	}

	@Override
	public void init() {
		super.init();

		// TODO: Acquire services if necessary
	}
	
	@Override
    public void processEvent(PluginEvent event) {
        if (event instanceof ProgramActivatedPluginEvent) {
        	provider.action.setEnabled(((ProgramActivatedPluginEvent)event).getActiveProgram() != null);
        }
    }

	// TODO: If provider is desired, it is recommended to move it to its own file
	private static class MyProvider extends ComponentProvider {

		private SymbolTransferPlugin plugin;
		private JPanel panel;
		private DockingAction action;

		public MyProvider(SymbolTransferPlugin plugin, String owner) {
			super(plugin.getTool(), owner, owner);
			this.plugin = plugin;
			buildPanel();
			createActions();
		}

		// Customize GUI
		private void buildPanel() {
			panel = new JPanel(new BorderLayout());
			JTextArea textArea = new JTextArea(5, 25);
			textArea.setEditable(false);
			panel.add(new JScrollPane(textArea));
			setVisible(true);
		}

		// TODO: Customize actions
		private void createActions() {
			action = new DockingAction("My Action", getName()) {
				@Override
				public void actionPerformed(ActionContext context) {
					Msg.showInfo(getClass(), panel, "Custom Action", "Hello!");
				}
			};
			action.setMenuBarData(new MenuData(new String[] { "File", "Hello World" }));
			action.setToolBarData(new ToolBarData(Icons.ADD_ICON, null));
			action.setEnabled(getProgram() != null);
			action.markHelpUnnecessary();
			dockingTool.addLocalAction(this, action);
		}

		@Override
		public JComponent getComponent() {
			return panel;
		}
		
		private Program getProgram() {

			ProgramManager pm = dockingTool.getService(ProgramManager.class);
			if (pm != null) {
				return pm.getCurrentProgram();
			}
			return null;
	    }
	}
}
```

I would expect an entry inside the "File" menu but there isn't one. Neither in the main window nor inside an opened project.

**Comment by ryanmkurtz** [comment]
Is your plugin visible and enabled in the `File -> Configure` menu?

**Comment by Pyrdacor** [comment]
It isn't listed there.

**Comment by ryanmkurtz** [comment, marked as answer]
try using `packageName = CorePluginPackage.NAME,`

**Reply by Pyrdacor** [reply]
Maybe this should be updated in the examples? Not sure why this works actually but it does. :)

**Comment by ryanmkurtz** [comment]
Also, make sure you do `File -> Configure` in the CodeBrowser tool and not the Front End.

**Comment by Pyrdacor** [comment]
> try using `packageName = CorePluginPackage.NAME,`

`CorePluginPackage` could not be resolved.

**Reply by ryanmkurtz** [reply]
Did you update your java imports?

**Comment by Pyrdacor** [comment]
Now it partially works. When opening a program a second Code Browser window is opened (which is not desired) where my toolbar icon is added. The main menu does not contain the "File" menu. But in the toolbar there is a new "File" entry with my "Hello World" entry (also not desired). Thanks for the help so far.

**Reply by ryanmkurtz** [reply]
Ok, sounds like you are getting there.

**Reply by Pyrdacor** [reply]
One last question. How can I add the toolbar icon and menu entry to the main code view window? I don't want to create a new window. Tried a bit with getOwner() and stuff but nothing worked yet.

**Reply by ryanmkurtz** [reply]
I believe you want something like `plugin.getTool().addAction()` instead of `plugin.getTool().addLocalAction()`

**Reply by Pyrdacor** [reply]
Perfect works like a charm. Thank you very much for your help. :)

**Reply by Pyrdacor** [reply]
I marked the packageName fix as the answer as it was basically the thing which made the actions appear in the first place.
